torchtitan/train.py

```diff
@@ -551,10 +551,9 @@ def close(self) -> None:
             logger.info("Created seed checkpoint")
         else:
             trainer.train()
-    finally:
-        if trainer:
-            trainer.close()
-
-        if torch.distributed.is_initialized():
-            torch.distributed.destroy_process_group()
-            logger.info("Process group destroyed.")
+    except Exception:
+        raise
+    else:
+        trainer.close()
+        torch.distributed.destroy_process_group()
```

Comment:
destroy_process_group() is causing the hang?  or  trainer.close()?

Becuase ideally calling destroy_process_group() itself would not hang, if it does that seems like another bug we should look into

Reply:
`destroy_process_group()`, I attached the py-spy result in the summary.

Reply:
Hmm. Cc @kwen2501 is that supposed to happen?

Reply:
`torchrun` should crash other ranks in case one rank crashed. It seems it failed to do so here.

Reply:
There is nothing a user program can do to sense remote crash (which is different than timeout). It requires an upper mechanism to intervene -- which is the role of job launcher/manager.

Reply:
@fegin can you confirm if other ranks are actually exiting? 

- if so, then this is a torchrun bug. torchrun should kill the program if some rank exits
- if not, then we should not care about destroy_process_group hanging on the 'good' ranks.  instead, we should care about why the exception rank did _not_ exit.

Reply:
@wconstab There is a misunderstanding. Only 4 ranks called into destroy_process_group. Other 4 ranks were calling FSDP allgather. The exception is not from destroy_process_group but from user/model code. The all 8 ranks hanged and then eventually timeout.

Reply:
All behaviors described above are expected --
there are mismatched collectives, so those programs ARE expected to hang somewhere, whether it is `destroy_process_group` or just `print(x)`. 
`destroy_process_group` is 100% a GPU sync point. It does wait for enqueued collectives to complete. The one that doesn't wait is `abort_process_group`.

Reply:
> eventually timeout

This is the right behavior upon collective mismatch.

Reply:
My q is, why does the user program mute the exception and not re-throw? Does it believe it is recoverable? In this case it seems not recoverable?

Reply:
> My q is, why does the user program mute the exception and not re-throw? Does it believe it is recoverable? In this case it does not seem so?

Good question, @kwen2501 , I had the same impression as @wconstab. I thought `destroy_process_group` is a purely local call. That's why I wrapped it in `finally:` and it didn't go wrong until this week when we were debugging CP issues.

This PR should be the right way to call `destroy_process_group()`.